pkg/distributor/receiver/shim_test.go

```diff
@@ -0,0 +1,202 @@
+package receiver
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+	"github.com/weaveworks/common/logging"
+	"go.opentelemetry.io/collector/model/otlpgrpc"
+	"go.opentelemetry.io/collector/model/pdata"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+
+	"github.com/grafana/loki/pkg/logproto"
+)
+
+func TestParseEntry(t *testing.T) {
+	pLog := pdata.NewLogRecord()
+	pLog.SetName("testName")
+	pLog.SetFlags(31)
+	pLog.SetSeverityNumber(1)
+	pLog.SetSeverityText("WARN LEVEL")
+	pLog.SetSpanID(pdata.NewSpanID([8]byte{1, 2}))
+	pLog.SetTraceID(pdata.NewTraceID([16]byte{1, 2, 3, 4}))
+
+	now := time.Now()
+	pLog.SetTimestamp(pdata.NewTimestampFromTime(now))
+
+	entry, err := parseEntry(pLog, "json")
+	if err != nil {
+		t.Fatal(err)
+	}
+	expexted := logproto.Entry{
+		Timestamp: now,
+		Line:      `{"severity_number":1,"severity_text":"WARN LEVEL","name":"testName","body":"","flags":31,"trace_id":"01020304000000000000000000000000","span_id":"0102000000000000"}`,
+	}
+	require.Equal(t, expexted.Line, (*entry).Line)
+	require.Equal(t, expexted.Timestamp.UnixMilli(), (*entry).Timestamp.UnixMilli())
+
+	entry, err = parseEntry(pLog, "logfmt")
+	if err != nil {
+		t.Fatal(err)
+	}
+	expextedLogfmt := logproto.Entry{
+		Timestamp: now,
+		Line:      `body= flags=31 name=testName severity_number=1 severity_text="WARN LEVEL" span_id=0102000000000000 trace_id=01020304000000000000000000000000`,
+	}
+	require.Equal(t, expextedLogfmt.Line, (*entry).Line)
+
+}
+
+func TestParseLabel(t *testing.T) {
+	resource := pdata.NewAttributeMap()
+	resource.Insert("app", pdata.NewAttributeValueString("testApp"))
+	attr := pdata.NewAttributeMap()
+	attr.Insert("level", pdata.NewAttributeValueString("ERROR"))
+	labels := parseLabel(resource, attr)
+
+	expexted := `{app="testApp", level="ERROR"}`
+	require.Equal(t, expexted, labels)
+}
+
+func TestParseLog(t *testing.T) {
```

Comment:
If you accept the suggestion of making the parser available on its own package, some tests from the Loki exporter for otelcol could be incorporated here, to ensure backward compatibility. Like this:

https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/fcadca185b39cd2b4beebb6ac4acabd2f498593f/exporter/lokiexporter/exporter_test.go#L474-L504

Reply:
agree